tests/test_codebuild/test_codebuild.py

```diff
@@ -0,0 +1,704 @@
+        name=name,
+        source=source,
+        artifacts=artifacts,
+        environment=environment,
+        serviceRole=service_role,
+    )
+    client.start_build(projectName=name)
+
+    history = client.list_builds_for_project(projectName=name)
+    response = client.batch_get_builds(ids=history["ids"])
+
+    response.should_not.be.none
+    response["builds"].should_not.be.none
+    response["builds"][0]["currentPhase"].should.equal("COMPLETED")
+    response["builds"][0]["buildNumber"].should.be.a(int)
+    response["builds"][0]["phases"].should_not.be.none
+    len(response["builds"][0]["phases"]).should.equal(11)
+
+
+@mock_codebuild
+def test_codebuild_batch_get_builds_2_projects():
+    client = boto3.client("codebuild", region_name="eu-central-1")
+
+    source = dict()
+    source["type"] = "S3"
+    # repository location for S3
+    source["location"] = "bucketname/path/file.zip"
+    artifacts = {"type": "NO_ARTIFACTS"}
+
+    environment = dict()
+    environment["type"] = "LINUX_CONTAINER"
+    environment["image"] = "contents_not_validated"
+    environment["computeType"] = "BUILD_GENERAL1_SMALL"
+    service_role = (
+        "arn:aws:iam::{0}:role/service-role/my-codebuild-service-role".format(
+            ACCOUNT_ID
+        )
+    )
+
+    client.create_project(
+        name="project-1",
+        source=source,
+        artifacts=artifacts,
+        environment=environment,
+        serviceRole=service_role,
+    )
+    client.start_build(projectName="project-1")
+
+    client.create_project(
+        name="project-2",
+        source=source,
+        artifacts=artifacts,
+        environment=environment,
+        serviceRole=service_role,
+    )
+    client.start_build(projectName="project-2")
+
+    response = client.list_builds()
+    response["ids"].should_not.be.empty
+
+    for build_id in response["ids"]:
+        try:
+            build_id.should.contain("project-1")
+        except AssertionError:
+            build_id.should.contain("project-2")
+
+    for metadata in client.batch_get_builds(ids=response["ids"])["builds"]:
+        metadata.should_not.be.none
+        try:
+            metadata["id"].should.contain("project-1")
+        except AssertionError:
+            metadata["id"].should.contain("project-2")
+
+
+@mock_codebuild
+def test_codebuild_batch_get_builds_invalid_build_id():
+    client = boto3.client("codebuild", region_name="eu-central-1")
+
+    with pytest.raises(client.exceptions.InvalidInputException) as err:
+        client.batch_get_builds(ids=["some_project{}".format(uuid1())])
+    err.value.response["Error"]["Code"].should.equal("InvalidInputException")
+
+
+@mock_codebuild
+def test_codebuild_batch_get_builds_empty_build_id():
+    client = boto3.client("codebuild", region_name="eu-central-1")
+
+    with pytest.raises(ParamValidationError) as err:
+        client.batch_get_builds(ids=[])
+        err.response["Error"]["Code"].should.equal("ParamValidationError")
```

Comment:
```suggestion
    err.response["Error"]["Code"].should.equal("ParamValidationError")
```
Wrong indentation - this line doesn't actually get executed if `client.batch_get_builds` throws an error

Reply:
Nice catch thanks! From what I can see `ParamValidation` err object doesn't even have a response property like the others and its `value` is the return message "Invalid length for parameter ids, value: 0, valid min length: 1". I tried it and it seems like this would be better for checking the error type:
```python
err.typename.should.equal("ParamValidationError")
```

Reply:
Hmm.. `err.value.response["Error"]` should always give you a dict with the `Code` and a `Message`. You could also try to use `with pytest.raises(ClientError)`, instead of ParamValidationError, as that as the overarching exception-class, and should always behave the same

Reply:
Tried a few things. `ClientError` isn't catching ParamValidationError 😬. I can see that `typename` has been used throughout the code to prove out the return of catching a ParamValidationError. There's a few cases where I can see something like `assert err["Code"] == "ParamValidationError"` but that pattern isn't working for codebuild, which like you say I'd imagine it would.

```python
# passes
with pytest.raises(ParamValidationError) as err:
    client.batch_get_builds(ids=[])
err.typename.should.equal("ParamValidationError")

# fail with botocore.exceptions.ParamValidationError
with pytest.raises(ClientError) as err:
    client.batch_get_builds(ids=[])
err.value.response["Error"]["Code"].should.equal("ParamValidationError")
```